subxt/src/runtime_api/runtime_client.rs

```diff
@@ -0,0 +1,66 @@
+// Copyright 2019-2022 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use super::runtime_types::RuntimeApi;
+
+use crate::{
+    client::OnlineClientT,
+    error::Error,
+    Config,
+};
+use derivative::Derivative;
+use std::{
+    future::Future,
+    marker::PhantomData,
+};
+
+/// Execute runtime API calls.
+#[derive(Derivative)]
+#[derivative(Clone(bound = "Client: Clone"))]
+pub struct RuntimeApiClient<T, Client> {
+    client: Client,
+    _marker: PhantomData<T>,
+}
+
+impl<T, Client> RuntimeApiClient<T, Client> {
+    /// Create a new [`RuntimeApiClient`]
+    pub fn new(client: Client) -> Self {
+        Self {
+            client,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<T, Client> RuntimeApiClient<T, Client>
+where
+    T: Config,
+    Client: OnlineClientT<T>,
+{
+    /// Obtain a runtime API at some block hash.
+    pub fn at(
+        &self,
+        block_hash: Option<T::Hash>,
+    ) -> impl Future<Output = Result<RuntimeApi<T, Client>, Error>> + Send + 'static {
+        // Clone and pass the client in like this so that we can explicitly
+        // return a Future that's Send + 'static, rather than tied to &self.
+        let client = self.client.clone();
+        async move {
+            // If block hash is not provided, get the hash
+            // for the latest block and use that.
+            let block_hash = match block_hash {
+                Some(hash) => hash,
+                None => {
+                    client
+                        .rpc()
+                        .block_hash(None)
+                        .await?
+                        .expect("didn't pass a block number; qed")
```

Comment:
I don't get this `expect` the rpc call could fail if the connection is closed, or am I misunderstanding this?

Reply:
are you sure it looks like we call https://github.com/paritytech/subxt/blob/master/subxt/src/rpc/rpc.rs#L220-#L227?

Reply:
For the runtime API, as well as storage, I mimicked the events API that we have in place.

Yep, in our use-case the substrate node must provide the `best_hash` of the chain. And in all cases, the RPC should return a concrete `Some` value for the best hash (may it be genesis). 

The client backend must always provide a value for the `.info()` from [here](https://github.com/paritytech/substrate/blob/a447f03727ffd84020682f3e0e4044d5d282d274/primitives/blockchain/src/backend.rs#L37-L38).

I'd imagine that the initial thought was around the lines: if the substrate node misbehaves for this RPC query, we should not trust it anymore to submit further RPC calls; and if the substrate behavior does indeed change we'll capture it in our testing scenarios.

Reply:
> are you sure it looks like we call https://github.com/paritytech/subxt/blob/master/subxt/src/rpc/rpc.rs#L220-#L227?

I was showing the Substrate impl.

But yeah, I think the reason is simply that a node shoiuld always be able to give back a "latest" block hash, and the option only exists because it may not be able to give a block hash back at some explicit block number.

Reply:
ah, I missed the `await?.expect(..)`.

got it then I get why you where talking about the substrate impl :)